Source/core/SharedBuffer.cpp

```diff
@@ -26,148 +26,38 @@ namespace Thunder {
-        }
-#elif defined(__APPLE__)
-
-        uint32_t timeLeft = waitTime;
-        int semResult;
-        while (((semResult = sem_trywait(_semaphore)) != 0) && timeLeft > 0) {
-            ::SleepMs(100);
-            if (timeLeft != Core::infinite) {
-                timeLeft -= (timeLeft > 100 ? 100 : timeLeft);
-            }
-        }
-        result = semResult == 0 ? Core::ERROR_NONE : Core::ERROR_TIMEDOUT;
-#else
-
-        struct timespec structTime = {0,0};
-
-        clock_gettime(CLOCK_MONOTONIC, &structTime);
-        structTime.tv_nsec += ((waitTime % 1000) * 1000 * 1000); /* remainder, milliseconds to nanoseconds */
-        structTime.tv_sec += (waitTime / 1000) + (structTime.tv_nsec / 1000000000); /* milliseconds to seconds */
-        structTime.tv_nsec = structTime.tv_nsec % 1000000000;
-
-        do {
-            if (sem_clockwait(_semaphore, CLOCK_MONOTONIC, &structTime) == 0) {
-                result = Core::ERROR_NONE;
-            }
-            else if ( errno == EINTR ) {
-                continue;
-            }
-            else if ( errno == ETIMEDOUT ) {
-                result = Core::ERROR_TIMEDOUT;
-            }
-            else {
-                ASSERT(false);
-            }
-            break;
-        } while (true);
-#endif
-        return (result);
-    }
-
     SharedBuffer::SharedBuffer(const TCHAR name[])
         : DataElementFile(name, File::USER_READ | File::USER_WRITE | File::SHAREABLE, 0)
         , _administrationBuffer((string(name) + ".admin"), File::USER_READ | File::USER_WRITE | File::SHAREABLE, 0)
         , _administration(reinterpret_cast<Administration*>(PointerAlign(_administrationBuffer.Buffer())))
-#ifdef __WINDOWS__
+    #ifdef __WINDOWS__
         , _producer((string(name) + ".producer").c_str())
         , _consumer((string(name) + ".consumer").c_str())
-#else
-        , _producer(&(_administration->_producer))
-        , _consumer(&(_administration->_consumer))
-#endif
+    #else
+        , _producer(reinterpret_cast<uint8_t*>(_administration) + sizeof(Administration))
+        , _consumer(reinterpret_cast<uint8_t*>(_administration) + sizeof(Administration) + SharedSemaphore::Size())
+    #endif
         , _customerAdministration(PointerAlign(&(reinterpret_cast<uint8_t*>(_administration)[sizeof(Administration)])))
     {
         Align<uint64_t>();
     }
     SharedBuffer::SharedBuffer(const TCHAR name[], const uint32_t mode, const uint32_t bufferSize, const uint16_t administratorSize)
         : DataElementFile(name, mode | File::SHAREABLE | File::CREATE, bufferSize)
-        , _administrationBuffer((string(name) + ".admin"), mode | File::SHAREABLE | File::CREATE, administratorSize + sizeof(Administration) + (2 * sizeof(void*)) + 8 /* Align buffer on 64 bits boundary */)
+        , _administrationBuffer((string(name) + ".admin"), mode | File::SHAREABLE | File::CREATE, sizeof(Administration) + (SharedSemaphore::Size() * 2) + administratorSize + 
```

Comment:
_administration may be moved by PointerAlign, I think we need to cater for it too so another + sizeof(void*) to be on the safe side
_customerAdministration is may also be adjusted by PointerAlign, so another (sizeof(void*)...